hoptimator-jdbc/src/main/java/com/linkedin/hoptimator/jdbc/HoptimatorDdlExecutor.java

```diff
@@ -0,0 +1,221 @@
+    viewPath.add(viewName);
+    ViewTableMacro viewTableMacro = ViewTable.viewMacro(schemaPlus, sql, schemaPath,
+        viewPath, false);
+    ViewTable viewTable = (ViewTable) viewTableMacro.apply(Collections.emptyList());
+    try {
+      ValidationService.validateOrThrow(viewTable, TranslatableTable.class);
+      if (create.getReplace()) {
+        DeploymentService.update(viewTable, ViewTable.class); 
+      } else {
+        DeploymentService.create(viewTable, ViewTable.class);
+      }
+      schemaPlus.add(viewName, viewTable);
+    } catch (Exception e) {
+      throw new RuntimeException("Cannot CREATE VIEW in " + schemaName + ": " + e.getMessage(), e);
+    }
+  }
+
+  // N.B. copy-pasted from Apache Calcite
+  /** Executes a {@code CREATE MATERIALIZED VIEW} command. */
+  public void execute(SqlCreateMaterializedView create,
+      CalcitePrepare.Context context) {
+    final Pair<CalciteSchema, String> pair = schema(context, true, create.name);
+    if (pair.left == null) {
+      throw SqlUtil.newContextException(create.name.getParserPosition(),  
+          RESOURCE.schemaNotFound(create.name.getSimple()));
+    }
+    if (pair.left.plus().getTable(pair.right) != null) {
+      // Materialized view exists.
+      if (!create.ifNotExists && !create.getReplace()) {
+        // They did not specify IF NOT EXISTS, so give error.
+        throw SqlUtil.newContextException(create.name.getParserPosition(),
+            RESOURCE.tableExists(pair.right));
+      }
+      if (create.getReplace()) {
+        pair.left.plus().removeTable(pair.right);
+      } else {
+        // nothing to do
+        return;
+      }
+    }
+
+    final SqlNode q = renameColumns(create.columnList, create.query);
+    final String sql = q.toSqlString(CalciteSqlDialect.DEFAULT).getSql();
+    final List<String> schemaPath = pair.left.path(null);
+
+    SchemaPlus schemaPlus = pair.left.plus();
+    String schemaName = schemaPlus.getName();
+    String viewName = pair.right;
+    List<String> viewPath = new ArrayList<>();
+    viewPath.addAll(schemaPath);
+    viewPath.add(viewName);
+    try {
+      if (!(pair.left.schema instanceof Database)) {
+        throw new RuntimeException(schemaName + " is not a physical database.");
+      }
+      String database = ((Database) pair.left.schema).databaseName();
+
+      // Table does not exist. Create it.
+      ViewTableMacro viewTableMacro = ViewTable.viewMacro(schemaPlus, sql, schemaPath,
+          viewPath, false);
+      MaterializedViewTable materializedViewTable = new MaterializedViewTable(viewTableMacro);
+      RelDataType rowType = materializedViewTable.getRowType(new SqlTypeFactoryImpl(
+          RelDataTypeSystem.DEFAULT));
+      MaterializedView hook = new MaterializedView(context, database, viewPath, rowType, sql,
+          Collections.emptyMap());  // TODO support CREATE ... WITH (options...)
+      ValidationService.validateOrThrow(hook, MaterializedView.class);
```

Comment:
Is this where schema validation comes into play? Is there a reason to check this if it isn't an `update`?

Reply:
Right. I guess our current validations are only for changes to existing tables/views, but the validation API is open-ended. You can validate anything, theoretically. This hook probably isn't doing anything yet, but we might want to drop in materialized view validation logic at some point.